canister/src/types.rs

```diff
@@ -124,7 +127,17 @@ impl Block {
     }
 
     pub fn block_hash(&self) -> BlockHash {
-        BlockHash::from(self.block.block_hash())
+        if self.block_hash.borrow().is_none() {
+            // Compute the block_hash as it wasn't computed already.
+            // `block.block_hash()` is an expensive call, so it's useful to cache.
+            let block_hash = BlockHash::from(self.block.block_hash());
+            self.block_hash.borrow_mut().replace(block_hash);
+        }
+
+        self.block_hash
+            .borrow()
+            .clone()
+            .expect("block hash must be available")
```

Comment:
I'm not sure this is possible given that it's a refcell and we're replacing it one of the branches. Do you know how this can be done?

Reply:
Yeah actually matching wouldn't work. But I think the below would do what we want

```
self.block_hash
            .borrow_mut()
            .get_or_insert(BlockHash::from(self.block.block_hash()))
            .clone()
```

only checked quickly that it compiles, I didn't run the benchmark.

Reply:
Ah, right, forgot about `get_or_insert`. We actually need to use `get_or_insert_with` to reap the performance benefits, otherwise we'd also end up computing the block hash every time this method is called.